pkg/git/nativeGit/repoCache.go

```diff
@@ -205,12 +205,22 @@ func (r *RepoCache) cleanRepo(repoName string) {
 	r.lock.Unlock()
 }
 
-func (r *RepoCache) InstanceForRead(repoName string) (instance *git.Repository, err error) {
-	return r.instanceForRead(repoName, false)
+func (r *RepoCache) PerformAction(repoName string, fn func(repo *git.Repository)) error {
+	repo, err := r.instanceForRead(repoName, false)
+	r.lock.Lock()
```

Comment:
We should have a lock per repo.

This way we are really choking the system.
- 100 user reading repos, they must wait for each other.
- With per repo locks, we only put users of the same repo at odds with each other